MHVLib/HTTP/MHVHttpService.m

```diff
@@ -0,0 +1,276 @@
+
+    [Logger write:[NSString stringWithFormat:@"Begin request #%li", (long)currentRequest]];
+
+    //TODO: Could return NSURLSessionTask to allow tasks to be cancelled; that works for requests, but not blob uploads
+    [[self.urlSession dataTaskWithRequest:(NSURLRequest *)request
+                        completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error)
+      {
+          NSInteger statusCode = ((NSHTTPURLResponse *)response).statusCode;
+          
+          [Logger write:[NSString stringWithFormat:@"Response for #%li has status code: %li (%0.4f seconds)",
+                         (long)currentRequest, (long)statusCode, [[NSDate date] timeIntervalSinceDate:startDate]]];
+          
+          if (error)
+          {
+              completion(nil, error);
+          }
+          else
+          {
+              completion([self responseFromData:data urlResponse:response], nil);
+          }
+      }] resume];
+}
+
+- (void)downloadFileWithUrl:(NSURL *)url
+                 toFilePath:(NSString *)path
+                 completion:(void (^)(NSError *_Nullable error))completion
+{
+    NSURLRequest *request = [self requestWithUrl:url data:nil];
+    
+    [[self.urlSession downloadTaskWithRequest:request
+                            completionHandler:^(NSURL * _Nullable location, NSURLResponse * _Nullable response, NSError * _Nullable error)
+      {
+          if (error)
+          {
+              completion(error);
+              return;
+          }
+          
+          NSError *fileError;
+          if (![[NSFileManager defaultManager] copyItemAtURL:location toURL:[NSURL fileURLWithPath:path] error:&fileError])
+          {
+              completion(fileError);
+          }
+          else
+          {
+              [[NSFileManager defaultManager] setAttributes:@{
+                                                              NSFileProtectionKey : NSFileProtectionComplete
+                                                              }
+                                               ofItemAtPath:path
+                                                      error:nil];
+              completion(nil);
+          }
+      }] resume];
+}
+
+- (void)uploadBlobSource:(id<MHVBlobSource>)blobSource
+                   toUrl:(NSURL *)url
+               chunkSize:(NSUInteger)chunkSize
+              completion:(void (^)(MHVHttpServiceResponse *_Nullable response, NSError *_Nullable error))completion
+{
+    [self uploadBlobSource:blobSource toUrl:url chunkOffset:0 chunkSize:chunkSize completion:completion];
+}
+
+- (void)uploadBlobSource:(id<MHVBlobSource>)blobSource
+                   toUrl:(NSURL *)url
+             chunkOffset:(NSUInteger)chunkOffset
+               chunkSize:(NSUInteger)chunkSize
+              completion:(void (^)(MHVHttpServiceResponse *_Nullable response, NSError *_Nullable error))completion
+{
+    MHVASSERT_PARAMETER([url.scheme isEqualToString:@"https"]);
```

Comment:
The MHVASSERT_PARAMETER() macro is used to assert that a parameter is not nil. You shouldn't use it to validate the state of an object. You can just use MHVASSERT()